docs/ROADMAP.md

```diff
@@ -47,11 +47,57 @@ concepts.
 ### 6. Function catalog
 
 All of the hydration and validation logic has been moved from the kpt binary 
-to functions allowing for flexibility and security.  This enables new 
+to functions allowing for flexibility and security. This enables new 
 scenarios like limiting the customization and validation to a subset of 
 allowed functions.  The function catalog has received additional functions, 
 examples and help. Please visit the [function catalog] for further information.
 
+## In Progress
+
+### 7. Resource filtering in `kpt fn` commands
+
+Users want to invoke a kpt function (imperatively and declaratively) on a subset of 
+resources in the package by selecting them on the basis of GVKNN, package-path, 
+file path etc. For example, set work-load identity annotation on all Kubernetes 
+Service Account resources in this package [Tracker](https://github.com/GoogleContainerTools/kpt/issues/2015).
+**Estimated release date:** Last week of September 2021.
+
+
+### 8. Integrate kpt with Cloud Code
+
+One of the major areas of investment is to integrate [Cloud Code](https://cloud.google.com/code) with kpt to provide 
+package authoring assistance. Users can author Kptfile and functionConfig files with
+features like auto-complete and error detection. This significantly improves the 
+discoverability of Kptfile schema, catalog functions and their functionConfigs.
+**Estimated release date:** Last week of September 2021 for Kptfile schema integration,
+function catalog integration will be released by mid of November 2021.
+
+### 9. Pre/Post processing step in `kpt fn render`
+
+We need a way to pre(/post) process inputs(/outputs) involved in customizing the 
+package or any of its sub-packages before(/after) invoking the function pipeline [Tracker](https://github.com/GoogleContainerTools/kpt/issues/2419).
+e.g. Use-case to propagate common setters values in the package hierarchy before
+rendering the package resources.
+**Estimated release date:** Mid of November 2021.
+
+### 10. Improve functionConfig UX
+
+Package publishers want to provide command driven instructions (semi-automated)
+to configure a package that makes the user-guide/examples easy to follow.
+e.g. Porcelain to set values in functionConfig or generate functionConfig.
+**Estimated release date:** End of October 2021.
+
+Package consumers should be able to detect configuration errors early before the
+package is assumed to be ready. e.g. Introduce validation for functionConfig.
+**Estimated release date:** End of November 2021.
+
+### 11. Improve Function Authoring Experience
+
+We need a rich ecosystem of third party functions. Users should be able to write 
+functions with custom logic very quickly using the tools they are familiar with. 
+So we are investing on making function authoring experience very easy. This is an
+ongoing effort. **Estimated completion date:** End of December 2021.
+
```

Comment:
We need to capture a few more efforts:

1. Function Catalog Schema Standardization to enable discovering of functions and their metadata e.g. functionConfigs, attributes such as mutator/validator etc. [Q3]
   Integration of schema will be covered in rest of the items listed above e.g. improving functionConfig UX

2. New functions:
- set-project-id [Done]
- render-helm-chart [Done]
- enable-gcp-services [Q3]
- apply-time-mutation [Q3]
- apply-patch [Q4]

3. OCI support in packaging and schema integrations [Q4]
4. kpt live enhancements [Q4]
     - Improver error handling when resources fail to reconcile
     - Early termination of fatal errors
     - Better support for output in json format
     - Lifecycle directive
5. Task oriented docs enhancements [Q4]

Reply:
I am still a bit skeptical about the sequencing for OCI, do we have customers in Q4 who will use and deploy workloads based on this?

Reply:
> do we have customers in Q4 who will use and deploy workloads based on this?
I don't think so. We have had some indirect asks, but not a direct one. This fits in "nice to have" category so far. I think it is safe to remove it from the list and keep a watch on the enhancement request (the issue) to get a sense in Q4.

Reply:
Thank you!  It does seem like OCI is an promising feature, but the timing of it is not certain.  Maybe the roadmap should have the "up next" category that has no clear deliverables and no clear timelines but allows people to chime in on those tasks.  I'd love to say "we are thinking about OCI, tell us how you use OCI in your enterprise!"

Reply:
I like the idea of `up next` section to say `This is on our radar, and we are thinking about it`. I think users will benefit from that transparency and help us gather more feedback from users as well.

Reply:
We can add these to our internal docs first, debate and then make public. Since this is live doc we can always keep adding stuff.

Reply:
Ack.

Reply:
I think we need to figure out how to get more of the process to be inclusive of the community outside of Google.  I think a playbook leveraging some CNCF learnings would be a good thing.